docs/how-to/craft/use-layouts.rst

```diff
@@ -0,0 +1,106 @@
+snaps.
+
+Layouts can help `hooks <https://snapcraft.io/docs/supported-snap-hooks>`_ access any
+executables they may require.
+
+
+Define a layout
+---------------
+
+Define a layout as a map from target paths to source paths. Each source path takes the
+shape of a layout definition.
+
+.. code-block:: yaml
+    :caption: snapcraft.yaml
+
+    layout:
+      <target-path>:
+        <layout>
+      <target-path>:
+        <layout>
+
+Set the target path to the . Consult the
+
+Layout definitions use special syntax. Select one of the layout types, then fill it in
+for ``<layout>``:
+
+- ``symlink: <source-path>`` creates a symbolic link
+- ``bind: <source-path>`` bind-mounts a directory
+- ``bind-file: <source-path>`` bind-mounts a file
+- ``tmpfs: <source-path>`` mounts a private temporary in-memory filesystem
+
+Set ``<source-path>`` to ``$SNAP``, ``$SNAP_DATA`` or ``$SNAP_COMMON``.
+
+The ``symlink`` layout type is recommended because it's the fastest. The ``bind*`` and
+``tmpfs`` types *significantly* increase the startup time of the app. Some apps,
+however, treat symlinks differently than regular files or directories, so you might
+need to use a bind mount for them.
+
+Some apps might treat symlinks differently than regular files or directories, so you may
+need to use a bind mount.
+
+As ``/lib`` and ``/run`` directories in Linux are symbolic links to ``/usr/lib`` and
+``/var/run`` respectively, they require separate exceptions to ensure certain locations,
+such as ``/lib/firmware``, can't be worked around.
+
+
+Examples
+--------
+
+As an example, consider an app you want to package that:
+
+- Stores all data in ``/var/lib/foo``
+- Has a configuration file in ``/etc/foo.conf``
+- Uses read-only data in ``/usr/share/foo``
+
+A layout map that allows such software to be used without snap-specific modifications
+would look like:
+
+.. code-block:: yaml
+    :caption: snapcraft.yaml
+
+    layout:
+      /var/lib/foo:
+        bind: $SNAP_DATA/var/lib/foo
+      /usr/share/foo:
+        bind: $SNAP/usr/share/foo
+      /etc/foo.conf:
+        bind-file: $SNAP_DATA/etc/foo.conf
+
+Here's a realistic example of a layout map that make the glibc library available to the
+snap at runtime:
+
+.. code-block:: yaml
+    :caption: snapcraft.yaml
+
+    layouts:
+      /usr/lib/x86_64-linux-gnu/libEGL.so:
+        symlink: $SNAP/usr/lib/x86_64-linux-gnu/libEGL.so
+      /usr/lib/x86_64-linux-gnu/libGL.so:
+        symlink: $SNAP/usr/lib/x86_64-linux-gnu/libGL.so
+
+
+Manage files in read-only locations
+-----------------------------------
+
+Layouts can create new directories and files even in read-only locations such as
+``/usr/share``.
+
+For example, assume ``/usr/share/foo`` *doesn't* exist in the base snap. The following
+layout map creates ``/usr/share/foo``, visible only to executing snap apps:
+
+.. code-block:: yaml
+    :caption: snapcraft.yaml
+
+    layout:
+      /usr/share/foo:
+        bind: $SNAP/usr/share/foo
```

Comment:
The example for this confuses me. Earlier we stated that a tmpfs layout is needed for read-only objects, but here we use the regular bind type.

Reply:
I wouldn't stress on the snapd internals, it makes choices inside for some of these code paths.

What happens is that since the target does not exist, snapd under the hood creates a tmpfs to create that directory in a mount namespace to have a valid mount target